PPM-300: Add Bus agnostic API as optional dependency to the builds

abelog opened the pull request:
## Description

This PR intention is to introduce new HLAPI lib (High Level API) which should which links all BAAPI (Bus Agnostic API) libs.
`hlapi` is yet optional for build, as we don't use it right now.
Tested for Dummy build on CI. Flag was added by 113da938feb78a3d520efd092b31f19f667e98be commit. [Job](https://gitlab.com/prpl-foundation/prplMesh/-/jobs/674038950) and [pipeline](https://gitlab.com/prpl-foundation/prplMesh/-/pipelines/174989701).

## Changes

- Add `Findamx?.cmake` files to locate each `amx(c/d/j...) lib.
- Add new `hlapi` lib and new flag `ENABLE_HLAPI` to build it.

## Note

As we don't need to build `hlapi` right now - `ENABLE_HLAPI` flag isn't required yet, so no changes to regular build procedure used by CI and developers is required. Build in CI was done just for verification.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

abelog committed:
cmake: add Findamxc.cmake file

Add Findamxc.cmake file to locate AMXC library of BAAPI and relevant header
files.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
cmake: add Findamxb.cmake file

Add Findamxb.cmake file to locate AMXB library of BAAPI and relevant
header files.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
cmake: add Findamxd.cmake file

Add Findamxd.cmake file to locate AMXD library of BAAPI and relevant
header files.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
cmake: add Findamxj.cmake file

Add Findamxj.cmake file to locate AMXJ library of BAAPI and relevant
header files.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
cmake: add Findamxo.cmake file

Add Findamxo.cmake file to locate AMXO library of BAAPI and relevant
header files.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
cmake: add Findamxp.cmake file

Add Findamxp.cmake file to locate AMXP library of BAAPI and relevant
header files.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

arnout commented on cmake/Findamxc.cmake:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+# SPDX-FileCopyrightText: 2019-2020 the prplMesh contributors (see AUTHORS.md)
```
```suggestion
# SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
```

abelog replied:
Thank you for paying attention! Applied. Resolving.

arnout commented on cmake/Findamxc.cmake:
```diff
@@ -0,0 +1,34 @@
+# This code is subject to the terms of the BSD+Patent license.
+# See LICENSE file for more details.
+
+find_library(AMXC_LIBRARY "libamxc.so" PATHS ${CMAKE_PREFIX_PATH}/usr/lib/)
```
I don't think we need PATHS here. `find_library(AMXC_LIBRARY "libamxc.so")` should work as well. The PATHS thing was introduced for a specific situation where there was some dependency that was installed in a non-default location, I think.

abelog replied:
Applied. Resolving.

arnout commented on cmake/Findamxc.cmake:
```diff
@@ -0,0 +1,34 @@
+find_path(AMXC_INCLUDE_DIRS
+    NAMES amxc_llist.h amxc_string_join.h amxc_common.h amxc_rbuffer.h amxc_variant_type.h amxc_variant.h amxc_aqueue.h amxc_timestamp.h amxc_astack.h amxc_hash.h amxc_array.h amxc.h amxc_lstack.h amxc_string.h amxc_htable.h amxc_lqueue.h amxc_string_split.h
+    PATH_SUFFIXES amxc
+)
```
You don't need to list all of them, typically `amxc.h` should be enough.

Also, typically you'd include it with `#include <amxc/amxc.h>` so you should search directly for that file.

```suggestion
find_path(AMXC_INCLUDE_DIRS NAMES amxc/amxc.h)
```

abelog replied:
Applied. Only thing, that `PATH_SUFFIXES` [already point to the subdirectory which should be checked](https://cmake.org/cmake/help/latest/command/find_path.html?highlight=find_path), so i am not sure that is worth to implicitly add `amxc/amxc.h`.

arnout commented on CMakeLists.txt:
```diff
@@ -14,6 +14,7 @@ option (BUILD_AGENT "Build EasyMesh agent" ON)
 option (BUILD_CONTROLLER "Build EasyMesh controller" ON)
 
 option(BUILD_SHARED_LIBS "Build shared libraries (.so) instead of static ones (.a)" ON)
+option (ENABLE_AMBIORIX "Include Bus agnostic API in the build" OFF)
```
As I mentioned orally:
```suggestion
option (ENABLE_HLAPI "Build the northbound (high-level) API" OFF)
```

Also please turn it on in at least one build in CI.

abelog replied:
Applied. Will resolve as commit to alter CI to check build with flag is added.

arnout reviewed (commented):
Since draft, not approving yet.

I gave comments to the first Findxxx.cmake but obviously they apply to all.

abelog committed:
cmake: create separate HLAPI library

Create new Hight Level API library, which should handle interactions
with different system buses.

Jira link: https://jira.prplfoundation.org/browse/PPM-300

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

RanRegev commented on framework/platform/CMakeLists.txt:
```diff
@@ -1 +1,4 @@
\ No newline at end of file
+add_subdirectory(bpl)
+if (ENABLE_AMBIORIX)
+    add_subdirectory(HLAPI)
```
Consistency is important.
Better name this sub directory with lower case letters
```suggestion
    add_subdirectory(hlapi)
```

abelog replied:
Applied! Resolving.

RanRegev commented on framework/platform/HLAPI/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+set(MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
+
+# Common Sources
+file(GLOB_RECURSE hlapi_sources ${MODULE_PATH}/*.c*)
```
Since even CMAKE developers are [recommending NOT to use GLOB](https://cmake.org/cmake/help/latest/command/file.html#filesystem) for picking up all the files to be compiled, and since I have some more reasons on top on their's, I think it might be a good opportunity to start explicitly add the files that are part of the library.
On the other hand, it will introduce inconsistency with other parts of the project that still use GLOB.
However, the reason to ignore the inconsistency, is that specifying the files explicitly does not require any special understanding, it is straightforward, no learning curve at all.

abelog replied:
Then let's start to get rid of GLOB. Suggestion applied. Resolving.

RanRegev commented on framework/platform/HLAPI/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+project(hlapi VERSION ${prplmesh_VERSION})
```
maybe preceding with `prplmesh_`?
```suggestion
project(prplmesh_hlapi VERSION ${prplmesh_VERSION})
```

RanRegev commented on framework/platform/HLAPI/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+target_include_directories(${PROJECT_NAME}
+    PUBLIC
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>
```
AFAIK: the binary directory should go to /lib
```suggestion
        $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/lib>
```

abelog replied:
Accepted. Resolving.

RanRegev reviewed (commented):
Maybe it is too early in the process, but I guess we don't want just `hlapi.h` file that represents the **entire** high-level interface of prplmesh.
The HLAPI should be designed and it would probably end up with 
```cpp
namespace prplmesh::hlapi { /*...*/ }
```
and few classes and/or free functions within this namespace.

abelog commented:
@Mergifyio rebase
